### ISSUE TITLE
HERMES issues and improvements

### ISSUE THREAD

**Issue by mrieser** [matsim-org/matsim-libs#1248]
I came across **some issues** with HERMES that I plan to fix soon when I find some time. Still, I wanted to document them here so others are aware of them:

1. schedules containing stops that are not used in any TransitRoute results in IndexOutOfBoundsExceptions, because Hermes sizes some ArrayList based on the actual number of used stops, instead of the highest `Id<TransitStop>.index()`. 
2. Hermes wrongly assumes that TransitRoute-Ids are globally unique. But in reality, TransitRoute-Ids only have to be unique per TransitLine, but not for the whole schedule. So there could be two or more transit routes with the same Id, if they are in different transit lines. This can be observed in the following (and probably other) places:  
    - `ScenarioImporter#initRoutesStations()` adds an ArrayList for each route, but then accesses this in several places with `transitRoute.getId().index()`, which leads to collisions (transit routes of different transit lines accessing the same data).
    - `ScenarioImporter#initRoutesStations()` contains code like `line_of_route = new int[Id.getNumberOfIds(TransitRoute.class)];` which will lead to problems in such cases.

While detecting these issues and looking at the source code of Hermes, I found some places where probably some **performance optimizations** could be performed:

3. ScenarioImporter:271

    ```java
    if (stopIds.contains(sid)) {
    } else {
      stopIds.add(sid);
    }
    ```
    Simplify to just `.add(sid)`, there's no need to call `.contains` on a set. It just leads to unnecessarily accessing the data-structure twice in some cases.

4. ScenarioImporter:599ff  
There are multiple (at least 3) calls to `stop_ids.get(stopidx)` which could be cached instead of doing multiple calls.

5. ScenarioImporter:557

    ```java
    ArrayList<Integer> stop_ids = this.route_stops_by_index.get(rid);
    ```
    This could probably use `int[]` instead of `ArrayList<Integer>`, which would be much more efficient (storing `int` instead of `Integer` uses much less memory, access might be a tiny bit faster due to less (un)boxing.

6. ScenarioImporter / Realm.agent_stops:

    ```java
	// Agents waiting in pt stations. Should be used as follows:
	// agent_stops.get(curr station id).get(line id).get(dst station id) -> queue of agents
	protected ArrayList<ArrayList<Map<Integer, ArrayDeque<Agent>>>> agent_stops;
    ```
    - ArrayList, 1 entry for each stop
    - Each entry is one ArrayList for the potential TransitLineIds
    - Each entry in that is a Map<DestinationStopId, ArrayDeque<Agent>>, initialized with an empty `HashMap` by default.
    - This is very expensive, as for every stop, there is the data structure to store agents for every possible line. But many stops are only served by a few lines. Basically, this creates a huge number of unused, empty HashMaps.
    - --> HashMaps should only be created when they are actually used, the other ArrayList entries could be left `null`.
    - --> See if there is a better data structure, e.g. some "sparse array list" for storing the TransitLine data. It might be more efficient to do 10 int-comparisons along an array to find the right entry in a short array, than to access some random high array index.
    - --> The "Sparse ArrayList" could actually also be a `Map<TransitLineId, Map<DestinationStopId, ArrayDeque<Agent>>>`. But HashMaps have a high memory overhead in Java and are not very performant, as they jump around in memory. MATSim already has an `ArrayMap`, a Map-implementation backed by two simple arrays (one for the keys, one for the values), which often is faster for small map sizes, as it only has to iterate over a very short, linearly available memory region to find the right entry. One could even create a custom `IntArrayMap` based on this, using primitive `int` as a key, for further optimization. Assuming that most stops are typically served by less than 10 lines, and most lines have at most 20 stops, this should fulfill the requirement that linear iteration in the array is faster than a HashMap lookup, so it should be faster and use less memory.

Some **other issues** I found:

7. ScenarioImporter / Realm.agent_stops:

    ```java
	// Agents waiting in pt stations. Should be used as follows:
	// agent_stops.get(curr station id).get(line id).get(dst station id) -> queue of agents
	protected ArrayList<ArrayList<Map<Integer, ArrayDeque<Agent>>>> agent_stops;
    ```

    This stores for each stop and each transit line the agents that want to enter a vehicle of that transit line at that stop, sorted by stop where the passengers want to exit the transit vehicle. When a transit vehicle arrives at a stop (``org.matsim.core.mobsim.hermes.Realm#processAgentStopDepart(...)``), agents are added to the transit vehicle in the order of the stations where the agents want to get off (e.g. first the agents that want to exit at the next stop, then the agents that want to exit at the 2nd next stop, etc). As long as a vehicle has enough capacity, this shouldn't matter. But if a pt vehicle has limited capacity, it matters, as agents are not added on a first-come, first-serve basis, but depending on where they want to get out. And it differs from how QSim handled it.
    I guess we could simplify the data structure to have an `ArrayDeque` for the transit line, instead of for each upcoming stop on the line. This would reduce the number of ArrayDeques and allow to add the agents in a strict first-come, first-serve order. Currently, `processAgentStopDepart(...)` uses the destination stop id when adding the passenger to the transit vehicle, but this should also be available by accessing the passengers plan-entry in that method. (Hmm, actually, only the destination stop id, but not the stop index relative to the transit route, is available in that method... and that stop index is actually needed)

8. **Bug:** If two stops are mapped to the same link and used directly after each other in a transit route, Hermes calculates a "speed between stops" of 0, resulting in `timerunning` being 21747483647 (=`Integer.MAX_VALUE`) in `ScenarioImporter.generateDeterministicVehicleTrip()`, leading to an `IndexOutOfBoundsException`.

9. **Bug:** `ScenarioImporter.generateAgents()` calls `generateAgent()` for both regular persons and for transit vehicles. `generateAgent()` calls `new Agent(...)`, which sets `this.passengersByStop = new ArrayList();`. `Agent.isTransitVehicle()` is implemented as `return this.passengersByStop == null`, which returns false for all agents/vehicles, even for transit vehicles. I'm not yet sure what the implications of this will be...
  This also results in every agent having `this.passengersByStop` filled with 256 empty ArrayLists. This wastes quite a lot of memory! Assuming an empty ArrayList consumes 32 bytes, then 256 empty ArrayLists consume 8KB, and 1 million agents thus use around 7.5 GB of RAM just for empty, never-used ArrayLists...

10. **Bug:** This is actually a bit related to **8**: if `config.hermes().endTime()` is set to a value smaller than the last planned arrival of a transit vehicle at a stop (e.g. having a schedule running until 28:00:00, but one wants to stop the simulation already at 24:00:00), it will result in the same `IndexOutOfBoundsException` as in Pt.8, as Hermes assumes there will be no later Events than the specified end time. Will need to catch this somehow and probably add those events all to the latest available timestamp.

11. **Bug:** The precalculated event-times for the deterministic simulation seem to be off in some cases:
    ```
	<event time="0.0" type="TransitDriverStarts" driverId="pt_50468_305-27" vehicleId="50468" transitLineId="****" transitRouteId="****" departureId="50468"  />
	<event time="0.0" type="PersonEntersVehicle" person="pt_50468_305-27" vehicle="50468"  />
	<event time="0.0" type="vehicle enters traffic" person="pt_50468_305-27" link="80019094" vehicle="50468" networkMode="pt" relativePosition="1.0"  />
	<event time="0.0" type="VehicleArrivesAtFacility" vehicle="50468" facility="16341" delay="0.0"  />
	<event time="2.0" type="VehicleDepartsAtFacility" vehicle="50468" facility="16341" delay="2.0"  />
	<event time="2.0" type="entered link" link="-80019093" vehicle="50468"  />
	<event time="25.0" type="left link" link="-80019093" vehicle="50468"  />
	<event time="28.0" type="entered link" link="-80019092" vehicle="50468"  />
	<event time="29.0" type="left link" link="-80019092" vehicle="50468"  />
	<event time="31.0" type="left link" link="-80009534" vehicle="50468"  />
	<event time="32.0" type="entered link" link="-80009534" vehicle="50468"  />
	<event time="38.0" type="left link" link="80019094" vehicle="50468"  />
	<event time="38.0" type="entered link" link="-80009532" vehicle="50468"  />
	<event time="38.0" type="VehicleArrivesAtFacility" vehicle="50468" facility="16342" delay="-2.0"  />
	<event time="40.0" type="VehicleDepartsAtFacility" vehicle="50468" facility="16342" delay="0.0"  />
    ```
    In this case, the vehicle starts at link `80019094` but then enters link `-80019093` at time 2.0, but leaves the original link `80019094` at time 38.0. This leads to problems further down, e.g. in the scoring function, as the driver might have already left the vehicle, when the vehicle continues to enter and leave links...

**Comment by neuma** [comment]
True, only the combination of lineId AND routeId is globally unique. In fact, quite often the routes of a line are numbered consecutively, i.e. all lines have a route with id 0 or 1.

**Comment by jfbischoff** [comment]
Nice finding...and nice to see someone else is using Hermes :-)
Indeed, our transit schedules have unique transit route IDs, so I never stumbled across the issue. 
Thanks for fixing @mrieser !

**Comment by mrieser** [comment]
@DanGraur , @rodrigo-bruno : Could I have your opinion on points 5 and especially 6 (see description above)? 6 would require a simple, custom data structure, but I think with an enormous potential for memory savings and even some potential for higher performance. I'm happy to implement it, but maybe I'm overlooking something where you have more experience?

**Comment by rodrigo-bruno** [comment]
@mrieser , I agree with all the proposed changes. Some comments:

1. Higher ids will lead to more memory being pre-allocated in data structures. It would be good to keep this max as short as possible;

2-5. Fully agree, makes sense!

6: Yes, having a tighter data structure would help but it is often a compromise for performance. A data structure often used in GraalVM to solve the same problem is the [EconomicMap](https://github.com/oracle/graal/blob/c5c5934e912166b712a8802374ff3479452e3ac5/sdk/src/org.graalvm.collections/src/org/graalvm/collections/EconomicMapImpl.java) (might be worth checking).

7. Yes, that would work, assuming you can insert in the middle of the queue.

**Comment by mrieser** [comment]
@rodrigo-bruno Thanks for the pointer to the `EconomicMap`. This also uses just an array internally, very similar to my [ArrayMap](https://github.com/matsim-org/matsim-libs/blob/69e8ab29dd9553b93703f8fe16e7b7e1dd3440b1/matsim/src/main/java/org/matsim/core/utils/collections/ArrayMap.java), except that I just keep increasing the array when more items are added instead of switching to a hash-based ordering as the EconomicMap does. As long as only a small number of items as stored in the map, they should be pretty comparable.

7. "assuming you can insert in the middle of the queue": Why would I want to insert in the middle of the queue? Currently, agents waiting at a stop are inserted in different queues, depending on where they want to get out again. I would simply replace this with a single queue, so they are always added to vehicles in a first-come first-serve order instead of sorted by where they want to get out. Within the transit vehicle, they are still sorted into separate lists for when they have to get out, the same way as they are now.

**Comment by rodrigo-bruno** [comment]
>  Why would I want to insert in the middle of the queue? Currently, agents waiting at a stop are inserted in different queues, depending on where they want to get out again. I would simply replace this with a single queue, so they are always added to vehicles in a first-come first-serve order instead of sorted by where they want to get out. Within the transit vehicle, they are still sorted into separate lists for when they have to get out, the same way as they are now.

Got it, you are only replacing the data structure at the station and not inside the vehicle. Then, no problem at all! :-)

**Comment by mrieser** [comment]
Fixed in #1274.

**Comment by DanGraur** [comment]
@mrieser, sorry for the late response. After following the discussion, the proposals seem sound, and the results speak for themselves. Great work on these these ideas, and the implementation itself!
